### PR TITLE
Initial support for Bazel

### PR DESCRIPTION
**Pull request by rlnx**
## This PR includes:
- Bazel support for the following parts of old DAAL:
  - Services and Data Management
  - `classifier`, `covariance`, `distributions`, `dtrees/forest/classification`, `dtrees/forest/regression`, `engines`, 
     `kernel_function`, `kmeans`, `k_nearest_neighbors`, `normalization/zscore`, `pca`, `qr`, `svd`, `svm`.
- Bazel support for the new interfaces
- Bazel support for oneAPI C++ examples.
- Dependency management for TBB, FPK (incl. FPK for GPU), read env variables or download from GitHub dependencies.
- Automatic CPU detection for one-CPU build.
- Only statically linked binaries are available.
- Support Linux, Intel, GNU and DPC++ compilers.

### Set compiler
```sh
export CC=icc # To set Intel Compiler (default)
export CC=gcc # To set GNU compiler
CC=gcc bazel build :release # Example without export
```

### Build and run examples/tests
To run one test and see PASSED/FAILED (all needed dependencies will be built automatically)
```sh
bazel test //<test_target>
```

To run one test and see output (all needed dependencies will be built automatically)
```sh
bazel run //<test_target>
```

To run an example (all needed dependencies will be built automatically)
```sh
bazel test //examples/oneapi/cpp:pca_cor_dense_batch
```

To run all examples (all needed dependencies will be built automatically)
```sh
bazel test //examples/oneapi/cpp:all
```

#### Linking modes for tests
- **dev** mode (default). Link tests against only required dependencies for the `<test_name>`
   ```sh
   bazel test //<test_name> --@config//:test_link_mode=dev  # This is default flag value
   ```
- **static** mode. Link tests against static libraries
   ```sh
   bazel test //<test_name> --@config//:test_link_mode=static
   ```

- **release_static** mode. Link tests against static libraries from `DAALROOT`. Useful in case we have separate oneDAL build and we want to run only tests.
   ```sh
   bazel test //<test_name> --@config//:test_link_mode=release_static
   ```

### Build binaries
```sh
bazel build //cpp/daal:core_static            # Build libonedal_core.a
bazel build //cpp/daal:thread_static          # Build libonedal_thread.a
bazel build //cpp/daal:sequential_static      # Build libonedal_sequential.a
bazel build //cpp/oneapi/dal:static           # Build libonedal.a
bazel build //cpp/oneapi/dal:static_dpc       # Build libonedal_dpc.a
```

### Build release
Creates release directory structure in `bazel-bin/release`, currently includes header files and binaries.
```sh
bazel build :release      # Build release structure 
```

### CODE REVIEW

**Comment by napetrov** [comment]
2 issues found by CodeFactor - extra space between #! at file start
tool_not_found.tpl.sh
merge_static_libs_lnx.tpl.sh

File: makefile [hunk 1]
```diff
@@ -502,7 +489,32 @@ $(WORKDIR.lib)/$(core_y):                   $(daaldep.ipp) $(daaldep.vml) $(daal
                                             $(if $(PLAT_is_win32e),$(CORE.srcdir)/export_win32e.def) \
                                             $(CORE.tmpdir_y)/$(core_y:%.$y=%_link.txt) ; $(LINK.DYNAMIC) ; $(LINK.DYNAMIC.POST)
 
-$(CORE.objs_a): $(CORE.tmpdir_a)/inc_a_folders.txt
+$(CORE.kernel_defines): | $(shell dirname $(CORE.kernel_defines))/.
+	echo "/* file: daal_kernel_defines_gen.h */" >> $@
+	echo "/*******************************************************************************" >> $@
+	echo "* Copyright 2014-2020 Intel Corporation" >> $@
```
Comment (napetrov) [inline makefile:495]: let's remove year here

**Comment by rlnx** [comment]
/intelci: run
set infra_branch=dev/risrafil-initial-bazel

**Comment by rlnx** [comment]
/intelci: run
set infra_branch=dev/napetrov_bazel_init

**Comment by napetrov** [comment]
Overall looks ok - the only remaining problem is linux EOLs in autogenerated files on windows

**Comment by rlnx** [comment]
/intelci: run
set infra_branch=dev/napetrov_bazel_init

**Comment by rlnx** [comment]
/intelci: run
set infra_branch=dev/napetrov_bazel_init

**Pull request merged**